ocaml/xapi/pkg_mgr.mli

```diff
@@ -0,0 +1,17 @@
+type t = Yum | Dnf
+
+type sub_cmd = Reposync | Repoconfig | Repoquery
+
+val active : unit -> t
+
+val pmgr_cmd : unit -> string
+
+val reposync_cmd : unit -> string
+
+val repoconfig_cmd : unit -> string
+
+val repoquery_cmd : unit -> string
+
+val plugins : unit -> string list
+
+val add_sub_cmd : sub_cmd -> string list -> string list
```

Comment:
I'm not sure about this interface design, I'd prefer the signatures here to be of type

`t -> string` instead taking `unit`, otherwise why bother having `t`?

Another part of this is that once we have this in place, it would be better to make `t` abstract and remove some leaky abstractions in repository.ml. For example, command name and parameters are tigtly coupled, make the code reflect this:
```ocaml
val repoquery_installed : t -> (string * list string)

let query_installed t =
  let fmt = get_repoquery_fmt t in
  let params =
    (["-a"; "--qf"; fmt]
    @
    match t with
    | Yum ->
        ["--pkgnarrow=installed"]
    | Dnf ->
        ["--installed"]
    )
    |> add_sub_cmd Repoquery
  in
  ((repoquery_cmd t),  params)

...

let get_installed_pkgs t =
  let cmd, params = Pkg_mgr.query_installed t in
  Helpers.call_script cmd params
  |> Astring.String.cuts ~sep:"\n"
  ...
```

Reply:
An alternative would be to create a signature for package managers, implement two (one for dnf, one for yum), and at runtime pick one of the two modules:

```ocaml
let get_pkg_mgr config : (module Package_manager.S) =
  match condition_on config with
   Yum -> (module Pkg_mgr.Yum)
   Dnf ->
    (module Pkg_mgr.Dnf)

module Pkg_mgr = (val get_pkg_mgr conf)
```

Reply:
@psafont  Thanks for the reviewing, 
About the interface, technically, `t` is required for the signatures, like `reposync_cmd`, 
However, the running type of `t` is detected inside the function, so the caller do NOT need to bother it.

I was also considering ask the Pkg_mgr module to handle all the command and args so the callers only need to decide which function is needed.  When I go through the codes, I found lots of commands and args are used. 
So instead of put all the command args in the module, I provide general command like `add_sub_cmd`, and ask the caller construct concrete parameters.     But if that is preferred, I can follow the suggestion, with agree that the command line and the parameters are tightly bound.

Both the two options make sense to me,  I will follow one of them.

Reply:
I see the first option as a middle way between the current one and the two modules. It isolates some of the concerns while not all, but it also should have less code churn. I'd be happy with that for the time being

Reply:
I think it will be easy to unit test them as well if the real `run` function could be replaced with something like generating some string outputs.

Reply:
@minglumlu do you mean put the `run` function inside `Pkg_mgr`? 
I prefer NOT, as we are now decouple the output handling, for example, in some cases, yum wants stderr while dnf wants stdout.

Reply:
The outputs of YUM and DNF may be different. Is there a chance to unify the outputs so that xapi can handle with same interfaces and data structures?

Reply:
If the differences between YUM and DNF are mainly on parameters, this would be fine.